### ISSUE TITLE
[PATCH] Win32 MSVC perlglob expansion in Encode

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12072]
Migrated from [rt.perl.org#112612](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112612) (status was 'resolved')

Searchable as RT112612$

**Comment by p5pRT** [comment]
### From @rurban


This is a bug report for perl from rurban@&#8203;cpanel\.net\,
generated with the help of perlbug 1\.39 running under perl 5\.15\.9\.

From a5ad7f26363aa02510e89471d9fc277dacafc259 Mon Sep 17 00&#8203;:00&#8203;:00 2001
From&#8203;: Reini Urban \<rurban@&#8203;x\-ray\.at>
Date&#8203;: Wed\, 25 Apr 2012 13&#8203;:21&#8203;:20 \-0500
Subject&#8203;: \[PATCH\] Win32 MSVC perlglob expansion in Encode

Win32 needs to expand glob via perlglob\. But this needs perl\.exe\, not miniperl\.exe


---
 cpan/Encode/Byte/Makefile\.PL   |    2 \+\-
 cpan/Encode/CN/Makefile\.PL     |    2 \+\-
 cpan/Encode/EBCDIC/Makefile\.PL |    2 \+\-
 cpan/Encode/JP/Makefile\.PL     |    2 \+\-
 cpan/Encode/KR/Makefile\.PL     |    2 \+\-
 cpan/Encode/Symbol/Makefile\.PL |    2 \+\-
 cpan/Encode/TW/Makefile\.PL     |    2 \+\-
 7 files changed\, 7 insertions\(\+\)\, 7 deletions\(\-\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/cpan/Encode/Byte/Makefile.PL b/cpan/Encode/Byte/Makefile.PL
index 60caf38..88807ac 100644
--- a/cpan/Encode/Byte/Makefile.PL
+++ b/cpan/Encode/Byte/Makefile.PL
@@ -180,7 +180,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q" -"O"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/CN/Makefile.PL b/cpan/Encode/CN/Makefile.PL
index 6d54404..36a1d7b 100644
--- a/cpan/Encode/CN/Makefile.PL
+++ b/cpan/Encode/CN/Makefile.PL
@@ -156,7 +156,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/EBCDIC/Makefile.PL b/cpan/Encode/EBCDIC/Makefile.PL
index 0e2a13c..82e557a 100644
--- a/cpan/Encode/EBCDIC/Makefile.PL
+++ b/cpan/Encode/EBCDIC/Makefile.PL
@@ -137,7 +137,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q" -"O"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/JP/Makefile.PL b/cpan/Encode/JP/Makefile.PL
index a75685e..15a7f18 100644
--- a/cpan/Encode/JP/Makefile.PL
+++ b/cpan/Encode/JP/Makefile.PL
@@ -156,7 +156,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/KR/Makefile.PL b/cpan/Encode/KR/Makefile.PL
index e95d039..063a0d2 100644
--- a/cpan/Encode/KR/Makefile.PL
+++ b/cpan/Encode/KR/Makefile.PL
@@ -154,7 +154,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/Symbol/Makefile.PL b/cpan/Encode/Symbol/Makefile.PL
index 23ca1f4..0fa8e8d 100644
--- a/cpan/Encode/Symbol/Makefile.PL
+++ b/cpan/Encode/Symbol/Makefile.PL
@@ -142,7 +142,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q" -"O"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
diff --git a/cpan/Encode/TW/Makefile.PL b/cpan/Encode/TW/Makefile.PL
index 99c94bf..b427f0e 100644
--- a/cpan/Encode/TW/Makefile.PL
+++ b/cpan/Encode/TW/Makefile.PL
@@ -152,7 +152,7 @@ sub postamble
     $plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
     my $ucopts = '-"Q"';
     $str .=  
-        qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
+        qq{\n\t\$(FULLPERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
     open (FILELIST, ">$table.fnm")
         || die "Could not open $table.fnm: $!";
     foreach my $file (@{$tables{$table}})
-- 

```
</details>
1.7.9



---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=high
&nbsp;    module=Encode


---
Site configuration information for perl 5\.15\.9&#8203;:

Summary of my perl5 \(revision 5 version 15 subversion 8\) configuration&#8203;:
&nbsp;  Derived from&#8203;: dbeb8947c09c5d0f027aba8bb6db953a7ae36235
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=5\.1\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags =' \-g \-O2 \-DDEBUGGING \-DWIN32  \-DPERL\_TEXTMODE\_SCRIPTS \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-fno\-strict\-aliasing \-mms\-bitfields'\,
&nbsp;    optimize='\-g \-O2 \-DDEBUGGING'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion='4\.6\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='long long'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags ='\-g \-L"c&#8203;:\\perl516\\5\.15\.9\\lib\\MSWin32\-x86\-multi\-thread\\CORE" \-L"C&#8203;:\\MinGW\\lib" \-Wl=\-\-enable\-auto\-import'
&nbsp;    libpth=C&#8203;:\\MinGW\\lib
&nbsp;    libs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
&nbsp;    perllibs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
&nbsp;    libc=\, so=dll\, useshrplib=true\, libperl=libperl515\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-mdll \-g \-L"c&#8203;:\\perl516\\5\.15\.9\\lib\\MSWin32\-x86\-multi\-thread\\CORE" \-L"C&#8203;:\\MinGW\\lib" \-Wl=\-\-enable\-auto\-import'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: DEBUGGING HAS\_TIMES HAVE\_INTERP\_INTERN MULTIPLICITY
&nbsp;                        PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_IMPLICIT\_SYS
&nbsp;                        PERL\_MALLOC\_WRAP PERL\_PRESERVE\_IVUV
&nbsp;                        PERL\_TRACK\_MEMPOOL PL\_OP\_SLAB\_ALLOC USE\_ITHREADS
&nbsp;                        USE\_LARGE\_FILES USE\_LOCALE USE\_LOCALE\_COLLATE
&nbsp;                        USE\_LOCALE\_CTYPE USE\_LOCALE\_NUMERIC USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF



---
@&#8203;INC for perl 5\.15\.9&#8203;:
&nbsp;    C&#8203;:/cygwin/usr/src/perl/build\-5\.15\.9d\-win32/lib
&nbsp;    \.



---
Environment for perl 5\.15\.9&#8203;:
&nbsp;    HOME=/home/rurban
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/rurban/bin&#8203;:/usr/bin&#8203;:/usr/local/bin&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/home/rurban/bin&#8203;:/cygdrive/c/WINDOWS/system32&#8203;:/cygdrive/c/WINDOWS
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_JSON\_BACKEND="JSON&#8203;::XS"
&nbsp;    PERL\_YAML\_BACKEND="YAML"
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @rjbs

I'm curious as to why this isn't affecting everyone building on Win32\.  I see a number of reports 
from Win32 smokers that do not have problems building Encode\.

Do you suppose it's because they have another perl already built in their %PATH%?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Apr 25 11&#8203;:30&#8203;:29 2012\, rurban@&#8203;cpanel\.net wrote&#8203;:
> 
> This is a bug report for perl from rurban@&#8203;cpanel\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.15\.9\.
> 
> From a5ad7f26363aa02510e89471d9fc277dacafc259 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Reini Urban \<rurban@&#8203;x\-ray\.at>
> Date&#8203;: Wed\, 25 Apr 2012 13&#8203;:21&#8203;:20 \-0500
> Subject&#8203;: \[PATCH\] Win32 MSVC perlglob expansion in Encode
> 
> Win32 needs to expand glob via perlglob\. But this needs perl\.exe\, not
> miniperl\.exe

If\, by perlglob\, you mean the perlglob\.exe built from win32/perlglob\.c\,
then your sentence seems to say the opposite of what I know to be the case\.

miniperl’s glob operator uses perlglob\.  The Win32 Makefiles are
supposed to make sure that perlglob is in the path before running
miniperl\.  Is that perhaps not the case?

The full perl doesn’t use perlglob\.  By forcing the use of perl\.exe\, you
are actually bypassing perlglob and using File&#8203;::Glob\, for the glob operator\.

By using FULLPERL in Encode’s Makefiles\, I think \(without testing it\)
that you are breaking the build on Unix\, where Encode \(like other XS
extensions\) is built before perl\.  I don’t know what order the build
happens on Windows\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Thu May 03 20&#8203;:22&#8203;:11 2012\, sprout wrote&#8203;:
> On Wed Apr 25 11&#8203;:30&#8203;:29 2012\, rurban@&#8203;cpanel\.net wrote&#8203;:
> > 
> > This is a bug report for perl from rurban@&#8203;cpanel\.net\,
> > generated with the help of perlbug 1\.39 running under perl 5\.15\.9\.
> > 
> > From a5ad7f26363aa02510e89471d9fc277dacafc259 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> > From&#8203;: Reini Urban \<rurban@&#8203;x\-ray\.at>
> > Date&#8203;: Wed\, 25 Apr 2012 13&#8203;:21&#8203;:20 \-0500
> > Subject&#8203;: \[PATCH\] Win32 MSVC perlglob expansion in Encode
> > 
> > Win32 needs to expand glob via perlglob\. But this needs perl\.exe\, not
> > miniperl\.exe
> 
> If\, by perlglob\, you mean the perlglob\.exe built from win32/perlglob\.c\,
> then your sentence seems to say the opposite of what I know to be the
case\.
> 
> miniperl’s glob operator uses perlglob\.  The Win32 Makefiles are
> supposed to make sure that perlglob is in the path before running
> miniperl\.  Is that perhaps not the case?
> 
> The full perl doesn’t use perlglob\.  By forcing the use of perl\.exe\, you
> are actually bypassing perlglob and using File&#8203;::Glob\, for the glob
operator\.
> 
> By using FULLPERL in Encode’s Makefiles\, I think \(without testing it\)
> that you are breaking the build on Unix\, where Encode \(like other XS
> extensions\) is built before perl\.  I don’t know what order the build
> happens on Windows\.
> 

The Windows build makes miniperl\.exe first\, then perlglob\.exe\, then
pure\-Perl modules\, then perl\.exe\, then XS extensions\.

My \(almost nightly\) VC10 smoker builds everything fine and most
definitely does not have perl\.exe \(or miniperl\.exe or perlglob\.exe\) in
its PATH to start with\. I recently also ran one smoke each with VC6
through VC9\, and they all built fine as well\.

Can we see an example of the output from whatever build failure is being
encountered?

**Comment by p5pRT** [comment]
### From @rurban

My windows is fresh\, so I guess it's a missing perlglob in the PATH\. I'll test it tomorrow\.

Ricardo SIGNES via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

>I'm curious as to why this isn't affecting everyone building on Win32\.  I see a number of reports 
>from Win32 smokers that do not have problems building Encode\.
>
>Do you suppose it's because they have another perl already built in their %PATH%?

**Comment by p5pRT** [comment]
### From @rurban

On Fri\, May 4\, 2012 at 2&#8203;:21 AM\, Steve Hay via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Thu May 03 20&#8203;:22&#8203;:11 2012\, sprout wrote&#8203;:
>> On Wed Apr 25 11&#8203;:30&#8203;:29 2012\, rurban@&#8203;cpanel\.net wrote&#8203;:
>> >
>> > This is a bug report for perl from rurban@&#8203;cpanel\.net\,
>> > generated with the help of perlbug 1\.39 running under perl 5\.15\.9\.
>> >
>> > From a5ad7f26363aa02510e89471d9fc277dacafc259 Mon Sep 17 00&#8203;:00&#8203;:00 2001
>> > From&#8203;: Reini Urban \<rurban@&#8203;x\-ray\.at>
>> > Date&#8203;: Wed\, 25 Apr 2012 13&#8203;:21&#8203;:20 \-0500
>> > Subject&#8203;: \[PATCH\] Win32 MSVC perlglob expansion in Encode
>> >
>> > Win32 needs to expand glob via perlglob\. But this needs perl\.exe\, not
>> > miniperl\.exe
>>
>> If\, by perlglob\, you mean the perlglob\.exe built from win32/perlglob\.c\,
>> then your sentence seems to say the opposite of what I know to be the
> case\.
>>
>> miniperl’s glob operator uses perlglob\.  The Win32 Makefiles are
>> supposed to make sure that perlglob is in the path before running
>> miniperl\.  Is that perhaps not the case?
>>
>> The full perl doesn’t use perlglob\.  By forcing the use of perl\.exe\, you
>> are actually bypassing perlglob and using File&#8203;::Glob\, for the glob
> operator\.

Yes\. I mixed that up\.
miniperl failed to call perlglob\.
Using perl would need an already built File&#8203;::Glob\, but this is not the case

perlglob can be called from the shell ok\.
So I guess miniperl restricted the path for some perlglob dll\.
But I did not use filemon \(strace for windows\) yet\, to check what
exactly failed\.

>> By using FULLPERL in Encode’s Makefiles\, I think \(without testing it\)
>> that you are breaking the build on Unix\, where Encode \(like other XS
>> extensions\) is built before perl\.  I don’t know what order the build
>> happens on Windows\.

Right\. Encode is before File&#8203;::Glob

> The Windows build makes miniperl\.exe first\, then perlglob\.exe\, then
> pure\-Perl modules\, then perl\.exe\, then XS extensions\.
>
> My \(almost nightly\) VC10 smoker builds everything fine and most
> definitely does not have perl\.exe \(or miniperl\.exe or perlglob\.exe\) in
> its PATH to start with\. I recently also ran one smoke each with VC6
> through VC9\, and they all built fine as well\.
>
> Can we see an example of the output from whatever build failure is being
> encountered?

I"m sorry\, but I will not come to it for the next two weeks\.
I'm getting married\.
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Apr 25 11&#8203;:30&#8203;:29 2012\, rurban@&#8203;cpanel\.net wrote&#8203;:
> 
> This is a bug report for perl from rurban@&#8203;cpanel\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.15\.9\.
> 
> From a5ad7f26363aa02510e89471d9fc277dacafc259 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Reini Urban \<rurban@&#8203;x\-ray\.at>
> Date&#8203;: Wed\, 25 Apr 2012 13&#8203;:21&#8203;:20 \-0500
> Subject&#8203;: \[PATCH\] Win32 MSVC perlglob expansion in Encode
> 
> Win32 needs to expand glob via perlglob\. But this needs perl\.exe\, not
> miniperl\.exe

I think the underlying issue has been fixed by commit 25018358\.  See bug
\#113798\.

If that is not the case\, please say so and I will re\-open the ticket\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
